BLE API devirtualization

pan- opened the pull request:
### Description

This submission replace runtime polymorphism by compile time polymorphism in BLE main APIs, generic and pal layer.  

Removing virtual functions allows the linker to remove functions that are not used by application code. Here are some numbers for the BLE layer on NRF52840 compiled with GCC in release mode. 

| Application                 | Original | Devirtualized | Gain  |
|-------------------------|----------|---------------|-------|
| BLE_BatteryLevel        | 106447   | 78307         | 28140 |
| BLE_Beacon              | 105975   | 75369         | 30606 |
| BLE_Button              | 106443   | 78303         | 28140 |
| BLE_GAP                 | 107112   | 79456         | 27656 |
| BLE_GAPButton           | 106673   | 76067         | 30606 |
| BLE_GattClient          | 106523   | 81433         | 25090 |
| BLE_GattServer          | 106089   | 77433         | 28656 |
| BLE_HeartRate           | 106477   | 78335         | 28142 |
| BLE_LED                 | 106443   | 77491         | 28952 |
| BLE_LEDBlinker          | 105800   | 81524         | 24276 |
| BLE_PeriodicAdvertising | 106794   | 81774         | 25020 |
| BLE_Privacy             | 106162   | 83000         | 23162 |
| BLE_SM                  | 106202   | 83364         | 22838 |
| BLE_Thermometer         | 106513   | 78373         | 28140 |
| TestApplication         | 109151   | 101335        | 7816  |
   
Even in our test application that uses all functions defined by the BLE API, we save around 7% of code. 

More importantly this construct is the foundation for efficient compile time configuration (With both optimisation, the beacon example shrink to ~30K of code).
 
#### Principle 

Compile time polymorphism is achieved with the [Curiously Recurring Template Pattern](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern) (CRTP). In a nutshell, every interface definition is now a class template that accepts its implementation as parameter. Interface APIs call the (hidden) implementation defined in the derived class.  If the function is not defined in the derived class then it fallbacks to a stub defined in the interface.

```c++
template <class Impl>
struct Interface {
    void foo() {
        // call to the implementation of foo
        static_cast<Impl*>(this)->foo_();
    }

protected:
    // convention is to use the suffix _ to name an implementation 
    void foo_() {
        printf("foo() default implementation");
    }
};

class Implementation : public Interface<Implementation> {
protected:
    void foo_() {
        printf("foo() derived implementation");
    }
};


class Stub : public Interface<Stub> { };

Implementation imp; 
Stub stub; 

imp.foo(); // call Implementation::foo_
stub.foo(); // call Interface::foo_
```

#### Backward compatibility   

User facing API prototypes are now templates defined in the namespace `ble::interface`: 
* `::Gap` -> `ble::interface::LegacyGap`
* `::ble::Gap` -> `ble::interface::Gap`
* `::GattClient` -> `ble::interface::GattClient`
* `::GattServer` -> `ble::interface::GattServer`
* `::SecurityManager` -> `ble::interface::SecurityManager`

It is expected that an implementation of ble forward the final implementation types in a header named `BleImplementationForward.h` and in the namespace `ble::impl`: 
* `ble::impl::LegacyGap` -> `ble::interface::LegacyGap<LegacyGapImpl>`
* `ble::impl::Gap` ->  `ble::interface::Gap<GapImpl>`
* `ble::impl::GattClient` -> `ble::interface::GattClient<GattClientImpl>`
* `ble::impl::GattServer` -> `ble::interface::GattServer<GattServerImpl>`
* `ble::impl::SecurityManager` -> `ble::interface::SecurityManager<SecurityManagerImpl>`

These final types are then imported into the global namespace to expose types similar to the one we've been using for years. Note that to not leak information from the implementation class it is important to export the interface and not the implementation: 

```c++ 
namespace ble {
namespace impl {

// Correct
typedef ble::interface::SecurityManager<SecurityManagerImpl> SecurityManager;

// Incorrect
//typedef SecurityManagerImpl SecurityManager;

} // impl  
} // ble 
```
##### Instantiation 

To prevent expensive instantiation of template types (and horrible inclusion pattern for the generic layer ... ), function body of the template interfaces are **not** defined in headers files. Instead these are defined in `.tpp` files. It is expected from an implementation to instantiate the types it export in a **cpp** file: 

```c++ 
#include "ble/GattClient.h"
#include "source/GattClient.tpp"
#include "GattClientImplementation.h" // include vendor implementation. 

// instantiate GattClient 
template class ble::interface::GattClient<GattClientImplementation>;
```

##### Out of scope 

The following items would improve this submission but are out of scope for now: 
* Devirtualization of ::BLEInstanceBase. 
* Extract non dependent types in interfaces.

##### TODO

   -  [ ] Squash history 

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [X] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@kjbracey-arm @paul-szczepanek-arm @donatieng 

<!-- 
    Optional
    Request additional reviewers with @username
-->

bridadan reviewed (approved):
Boy, that was an easy tools review!

Be sure to add this to the list of supported file types in the docs (you can follow my similar pr here: https://github.com/ARMmbed/mbed-os-5-docs/pull/952).

cmonr commented on tools/resources/__init__.py:
```diff
@@ -413,6 +413,7 @@ def add_directory(
         ".h": FileType.HEADER,
         ".hh": FileType.HEADER,
         ".hpp": FileType.HEADER,
+        ".tpp": FileType.HEADER,
```
I remember a day when header files ended with a `.h`, and we liked it.

pan- replied:
Do you ? Headers of the C++ library, a 35 years old language, don't have extensions. In C, X macros and friends have been traditionally put in files that don't end with `.h` too. 
If the `.h` extension as not been standardised at the language level it is because other extensions predate standardisation and these use cases are legitimate.

kjbracey replied:
Counter-pedant alert, and primed for counter-counter-pedantry.

Technically, the C++ standard doesn't say anything about file extensions.  The standard headers are included via `#include <atomic>` or whatever in the C++ source, but they could perfectly well be stored on disc as `atomic.hh` or `iso_atomic.hpp` or just built in to the compiler. Mechanism could be totally separate to whatever the `#include "xxx"` does for user headers. (Similarly `#include "myheader.h"` might not actually find the file `myheader.h`. On RISC OS it would find `h.myheader`, with `.` being the directory separator).

I actually find it a bit annoying that vendors chose to drop suffixes from the C++ `std` filenames. Having no extension at all is quite unhelpful for most other tools.

Anyway, no strong feelings about the separate prefix here. I also prefer to have special prefixes for things that you only want `#include`d with proper care, but it's up to tools folks to object. This one's a bit borderline, because nothing would actually break if randomly included, right? Would just slow down the build.

0xc0170 replied:
@pan- There's PR to add `.inc` files, wouldnt this extension cover the use case for `.tpp` ?

pan- replied:
> This one's a bit borderline, because nothing would actually break if randomly included, right?

@kjbracey-arm There's no header guards `.tpp` files at the moment. With header guards nothing would break.

@0xc0170 I choose `tpp` as an extension because it is more descriptive than `h` or `hpp` (and avoid clashes with actual header files). Any extension can be use. What is expected in `inc` files ?

0xc0170 replied:
I don't think we documented this - only extensions and purpose, without "what should be there". 

`inc` are used currently for some autogenerated headers in TFM code.

It might be better if this extension addition is sent via separate PR where we can discuss this and alos update docs (see https://github.com/ARMmbed/mbed-os/pull/9715)

bridadan replied:
From a tools perspective, I have no preference on the extensions. We can adapt to what you need.

The `.inc` extension is being added because the TF-M work is importing an external project into Mbed OS and the external project use `.inc` files. So we're simplifying the import process for that team.

sg- replied:
Does *.tpp work with other builders ie: MDK, DS-5, IAR, etc... or will this break our exporters? How is wrapping this in pragma once or ifdef... not good enough to keep the code base consistent? Apologies if this was already considered, the thread is quite long and I didn't read it all.

pan- replied:
I haven't tried for all exporters (but I can!). In this case, `*.tpp` files embed template class implementations and these are instantiated only once in a cpp file dedicated to the instantiation of the implementation type that contains all the required implementations. 

These files can't be named `.cpp` as these are not compilation unit and cannot used the sole suffix `.h` as there already is a header containing the declaration so we would end up with two headers with the same name. 

Suffix for template implementation really is just a convention as long as it respects the two requirements mentioned above. I started out with the suffix `.impl.h` and am happy to change the suffix to whatever is decided. 

----

Note that it would be possible to include everything in header files (I tried) at the expense of weird include patterns, longer compilation time and random breakage requiring reordering if things get moved around. So instead I tried to keep template implementation files consistent with the way we write source implementation files (.cpp) beside the name.

sg- replied:
Thanks @pan- it would be good to confirm the exporters are working as well as extend the tests for exporters to exercise the file type support to match what the os is capable of building. @bridadan

paul-szczepanek-arm commented on features/FEATURE_BLE/targets/TARGET_CORDIO/BleImplementationForward.h:
```diff
@@ -0,0 +1,164 @@
+        typedef generic::GenericSecurityManager<
+            pal::vendor::cordio::CordioSecurityManager,
+            vendor::cordio::SigningEventMonitor
+        > GenericSecurityManagerImpl;
```
why both namespace "impl" and "Impl" suffix

pan- replied:
I've added some information in the description; basically there is implementation classes and implementation interfaces. In that case `GenericSecurityManagerImpl` is the implementation class. The interface imported in the global namespace will be `ble::interface::SecurityManager<GenericSecurityManagerImpl>`. We don't want to import `GenericSecurityManagerImpl` as it requires visibility of `GenericSecurityManagerImpl` and may export thing differently than `ble::interface::SecurityManager` (visibility of function can change as an example).

paul-szczepanek-arm commented on features/FEATURE_BLE/source/generic/GenericGattClient.tpp:
```diff
@@ -1086,20 +1105,21 @@ ble_error_t GenericGattClient::write(
              encryption == link_encryption_t::ENCRYPTED_WITH_SC_AND_MITM)
         ) {
-            cmd = GattClient::GATT_OP_WRITE_CMD;
+			cmd = Base::GATT_OP_WRITE_CMD;
```
someone still hasn't configured their new IDE ;)

paul-szczepanek-arm commented on features/FEATURE_BLE/source/generic/GenericSecurityManager.tpp:
```diff
@@ -34,18 +34,19 @@ namespace generic {
     bool mitm,
     SecurityIOCapabilities_t iocaps,
-    const Passkey_t passkey,
+    const uint8_t* passkey,
```
why?

kjbracey reviewed (commented):
Approve of the general concept. Makes sense as an optimisation on the assumption that you want to optimise for the "1 BLE implementation in the system" case.

Would presumably be a pessimisation with two implementations, but anyone putting two implementations in one image in these sorts of systems probably needs to rethink their ROM-saving approach. (Or am I wrong?)

> Here are some numbers for the BLE layer on NRF52840 compiled with GCC in release mode.

This is a great chance to do some performance comparison - I'd love to see the same numbers for all the other toolchains.  I believe GCC (and ARMC6?) stand to gain the most by this, as they can't eliminate unused virtuals. Presumably IAR and ARMC5 gain much less, as they should be eliminating unused virtuals.

I believe the same CRTP approach may be worthwhile on a smaller scale for a bunch of our HAL driver classes, with their `virtual void lock()` functions.  Probably not many images are trying to simultaneously use `SPI` and `UnlockedSPI : public SPI`, and they certainly don't need run-time polymorphism.

(Although some key things do need it, like `FileHandle` or `Socket`, so we should still be looking at how to get compilers to optimise virtuals when actually required - eg spotting that `UARTSerial` is the only `FileHandle` in the build as LTO?).

pan- committed:
BLE - add static interface helper.

pan- committed:
BLE - replace some usages of Gap::Handle_t by ble::connection_handle_t

pan- committed:
BLE - devirtualize pal::AttClient

pan- committed:
BLE - devirtualize pal::AttClientToGattClientAdapter

pan- committed:
BLE - Devirtualize pal::ConnectionEventMonitor

The event handler has been extracted out of the monitor declaration.

pan- committed:
BLE - Devirtualize pal::Gap

The event handler has been taken out of Gap declaration and the instantiation must provide an implementation and the type that plays the event handler role.

pan- committed:
BLE - Devirtualize pal::GattClient

The event handler has been taken out of GattClient declaration and an instantiation requires the actual implementation and the type that handle events.

pan- committed:
BLE - Devirtualize pal::SecurityManager

The event handler has been extracted out of SecurityManager declaration and instantion of the interface requires the implementation and event handler type.

pan- committed:
BLE - Devirtualize pal::SigningEventMonitor

The event handler has been extracted out of SigningEventMonitor declaration and SigningEventMonitor instantion requires the implementation and event handler type.

pan- committed:
BLE - Devirtualization of ::Gap

The interface definition now lives in ::ble::interface::LegacyGap. Implementation must export the implementation type as ble::impl::LegacyGap.

pan- committed:
BLE - Devirtualize ::ble::Gap

The interface definition now lives in ::ble::interface::Gap.
The implementation must export the implementation types as ::ble::impl::Gap.

pan- committed:
BLE - Devirtualize ::GattClient

Interface definition now lives in ble::interface::GattClient. An implementation must export the implementation type in ::ble::impl::GattClient.

pan- committed:
BLE - Devirtualize ::GattServer.

The interface is defined in ::ble::interface::GattServer and an implementation must export the implementation type ::ble::impl::GattServer.

pan- committed:
BLE - Devirtualize ::SecurityManager

The interface now lives in ::ble::interface::SecurityManager. The implementation type is expectected to exported as ble ::ble::impl::SecurityManager by the implementation.

pan- committed:
BLE - Cleanup deprecation implementation bits in BLE.cpp

pan- committed:
BLE - cleanup includes.

pan- committed:
BLE - Devirtualize ble::generic::GenericGap

Expected types are similar to the type expected by the constructor:
- PalGap
- PalSecurityManager
- ConnectionEventMonitorEventHandler

Note that for the PalGap we expect a **template** of the form PalGap<EventHandler>

pan- committed:
BLE - Devirtualize ble::generic::GattClient

GenericGattClient is parametized by two types:
- The template of the PalGattClient
- The SigningMonitorEventHandler

Note that the PalGattClient template must be of the form PalGattClient<EventHandler>. The event handler being the GenericGattClient.

pan- committed:
BLE - Devirtualize ::ble::generic::GenericSecurityManager

Two parameters are expected by the template:
- TPalSecurityManager a template class of the security manager of the form TPalSecurityManager<EventHandler>. The GenericSecurityManager is the event handler.
- SigningMonitor a template in the form SigningMonitor<Handler>.

pan- committed:
BLE - Devirtualization of the Cordio port.

pan- committed:
BLE - Devirtualization of the Maxim port .

pan- committed:
BLE - Devirtualization of the legacy NRF51822 port.

pan- committed:
BLE - Devirtualization of the NRF51 port.

pan- committed:
BLE - Devirtualization of the NRF52 softdevice port.

pan- committed:
BLE - accept tpp extensions for header files.

pan- committed:
BLE - Fix qualification of abort and terminate.

donatieng reviewed (approved):
Looks really good - very excited about seeing this one coming in. I think we should discuss how we could apply the same solution to other areas of the OS.
